### ISSUE TITLE
Add partition_time function?

### ISSUE THREAD

**Issue by mrocklin** [pytoolz/toolz#290]
Thoughts on a function that partitioned an input iterator by a time interval?  This would be useful to batch a stream of data coming off of some external data source like a feed or database connection.

### Example

```python
def slowinc(x):
    sleep(0.1)
    return x + 1

>>> a = map(slowinc, range(100))
>>> b = partition_time(0.5, a)
>>> next(b)
(1, 2, 3, 4, 5)
>>> next(b)
(6, 7, 8, 9, 10)
```

This would require the use of a separate thread.

**Comment by mrucci** [comment]
Independently from the implementation, an additional parameter to bound the size of the returned sequence could be helpful, like in [riemann.streams/batch][batch].

[batch]: http://riemann.io/api/riemann.streams.html#var-batch

**Comment by mrocklin** [comment]
That's a very cool project; thank you for the pointer.  I wonder if there are other similar functions that would also be helpful to add.

**Comment by eriknw** [comment]
Ooh, I need to learn more about riemann.  Thanks!

So far, `toolz` has not entered the time domain.  This is potentially useful, and I suspect there are other functions that we may want.  I wonder how useful these would be however.  Asynchronous or concurrent frameworks often have tools or extensions to manage things like this.  Still, I could get behind a minimal set of functions that operate in the time-domain without the need of an asynchronous framework.

**Comment by mrocklin** [comment]
I agree that it would be good to establish scope here.  I think that for as long as we're still operating on iterators we're possibly still good for toolz inclusion.  If operations are better handled through some first-class stream or channel object then we should probably fork off to another library or, as you suggest, leave this work for others.

**Comment by mrocklin** [comment]
Regardless, the prototyping should probably happen externally before we consider inclusion.
